Importamos librerias:

In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
%matplotlib inline
import seaborn as sns

In [2]:
#cargamos el csv
df_hosp = pd.read_csv("dataset\hospitalizaciones_train.csv")

Comenzamos a analizar:  

**Available Extra Rooms in Hospital:** Habitaciones adicionales disponibles en el hospital. Una habitación no es igual a un paciente, pueden ser individuales o compartidas.  

**Department:** Área de atención a la que ingresa el paciente.  

**Ward_Facility_Code:** Código de la habitación del paciente.  

**doctor_name:** Nombre de el/la doctor/a a cargo del paciente.  

**staff_available:** Cantidad de personal disponible al momento del ingreso del paciente.  

**patientid:** Identificador del paciente.  

**Age:** Edad del paciente.  

**gender:** Género del paciente.  

**Type of Admission:** Tipo de ingreso registrado según la situación de ingreso del paciente.  

**Severity of Illness:** Gravedad de la enfermedad/condición/estado del paciente al momento del ingreso.  

**health_conditions:** Condiciones de salud del paciente.  

**Visitors with Patient:** Cantidad de visitantes registrados para el paciente.  

**Insurance:** Indica si la persona posee o no seguro de salud.  

**Admission_Deposit:** Pago realizado a nombre del paciente, con el fin de cubrir los costos iniciales de internación.  

**Stay (in days):** Días registrados de estancia hospitalaria.   

In [3]:
df_hosp.head()

Available Extra Rooms in Hospital  Department Ward_Facility_Code  \
0                                  4  gynecology                  D   
1                                  4  gynecology                  B   
2                                  2  gynecology                  B   
3                                  4  gynecology                  D   
4                                  2  anesthesia                  E   

  doctor_name  staff_available  patientid    Age  gender Type of Admission  \
0   Dr Sophia                0      33070  41-50  Female            Trauma   
1   Dr Sophia                2      34808  31-40  Female            Trauma   
2   Dr Sophia                8      44577  21-30  Female            Trauma   
3   Dr Olivia                7       3695  31-40  Female            Urgent   
4     Dr Mark               10     108956  71-80    Male            Trauma   

  Severity of Illness health_conditions  Visitors with Patient Insurance  \
0             Extreme          Diabetes                      4       Yes   
1               Minor     Heart disease                      2        No   
2             Extreme          Diabetes                      2       Yes   
3            Moderate              None                      4        No   
4            Moderate          Diabetes                      2        No   

   Admission_Deposit  Stay (in days)  
0        2966.408696               8  
1        3554.835677               9  
2        5624.733654               7  
3        4814.149231               8  
4        5169.269637              34

In [4]:
#pasamos la columna de dias a mayor a 8 dias =1 o ,menor a 8 dias = 0 
df_hosp["pred"] = df_hosp["Stay (in days)"].apply(lambda x: 1 if x > 8 else 0)

In [5]:
#dropeamos la columna "Stay (in days)" ya que es redundante.
df_hosp = df_hosp.drop(["Stay (in days)"],axis=1)

#### EDA VARIABLES CATEGÓRICAS / NUMERICAS

Control de correlacion de las columnas con respecto a "pred"  

Para calcular la correlación entre datos numéricos y categóricos en Python, puedes usar la función pandas.crosstab() junto con la función scipy.stats.chi2_contingency(). Primero, crea una tabla de contingencia usando la función pandas.crosstab() y luego calcula la correlación entre las variables utilizando scipy.stats.chi2_contingency()

La función stats.chi2_contingency() devuelve una tupla de cuatro valores. Los dos primeros valores son el valor del estadístico chi-cuadrado y el p-valor correspondiente. El tercer valor es el número de grados de libertad asociados con el estadístico chi-cuadrado. El cuarto valor es una matriz de contingencia que contiene los valores observados en la tabla de contingencia.  

En mi caso usare P para ver la relacion entre columnas

In [6]:
#traemos la libreria scipy
from scipy import stats

In [7]:
# Crea una tabla de contingencia
cor= pd.crosstab(df_hosp["pred"], df_hosp["Available Extra Rooms in Hospital"])
# Calcula la correlación entre las variables
chi2, p, dof, expected = stats.chi2_contingency(cor)
print(p)

0.0


In [8]:
# Crea una tabla de contingencia
cor= pd.crosstab(df_hosp["pred"], df_hosp["Department"])
# Calcula la correlación entre las variables
chi2, p, dof, expected = stats.chi2_contingency(cor)
print(p)

0.0


In [9]:
# Crea una tabla de contingencia
cor= pd.crosstab(df_hosp["pred"], df_hosp["Ward_Facility_Code"])
# Calcula la correlación entre las variables
chi2, p, dof, expected = stats.chi2_contingency(cor)
print(p)

0.0


In [10]:
# Crea una tabla de contingencia
cor= pd.crosstab(df_hosp["pred"], df_hosp["doctor_name"])
# Calcula la correlación entre las variables
chi2, p, dof, expected = stats.chi2_contingency(cor)
print(p)

0.0


In [11]:
# Crea una tabla de contingencia
cor= pd.crosstab(df_hosp["pred"], df_hosp["staff_available"])
# Calcula la correlación entre las variables
chi2, p, dof, expected = stats.chi2_contingency(cor)
print(p)

0.027185032657605395


In [12]:
# Crea una tabla de contingencia
cor= pd.crosstab(df_hosp["pred"], df_hosp["patientid"])
# Calcula la correlación entre las variables
chi2, p, dof, expected = stats.chi2_contingency(cor)
print(p)

0.2915705642154938


In [13]:
# Crea una tabla de contingencia
cor= pd.crosstab(df_hosp["pred"], df_hosp["Age"])
# Calcula la correlación entre las variables
chi2, p, dof, expected = stats.chi2_contingency(cor)
print(p)

0.0


In [14]:
# Crea una tabla de contingencia
cor= pd.crosstab(df_hosp["pred"], df_hosp["gender"])
# Calcula la correlación entre las variables
chi2, p, dof, expected = stats.chi2_contingency(cor)
print(p)

0.0


In [15]:
# Crea una tabla de contingencia
cor= pd.crosstab(df_hosp["pred"], df_hosp["Type of Admission"])
# Calcula la correlación entre las variables
chi2, p, dof, expected = stats.chi2_contingency(cor)
print(p)

1.6947748596830627e-21


In [16]:
# Crea una tabla de contingencia
cor= pd.crosstab(df_hosp["pred"], df_hosp["Severity of Illness"])
# Calcula la correlación entre las variables
chi2, p, dof, expected = stats.chi2_contingency(cor)
print(p)

2.8750137241379228e-130


In [17]:
# Crea una tabla de contingencia
cor= pd.crosstab(df_hosp["pred"], df_hosp["health_conditions"])
# Calcula la correlación entre las variables
chi2, p, dof, expected = stats.chi2_contingency(cor)
print(p)

5.658483183151773e-09


In [18]:
# Crea una tabla de contingencia
cor= pd.crosstab(df_hosp["pred"], df_hosp["Visitors with Patient"])
# Calcula la correlación entre las variables
chi2, p, dof, expected = stats.chi2_contingency(cor)
print(p)

0.0


In [19]:
# Crea una tabla de contingencia
cor= pd.crosstab(df_hosp["pred"], df_hosp["Insurance"])
# Calcula la correlación entre las variables
chi2, p, dof, expected = stats.chi2_contingency(cor)
print(p)

0.4988221951889673


In [20]:
# Crea una tabla de contingencia
cor= pd.crosstab(df_hosp["pred"], df_hosp["Admission_Deposit"])
# Calcula la correlación entre las variables
chi2, p, dof, expected = stats.chi2_contingency(cor)
print(p)

0.49327701140734725


#### Conclusion:  
Luego de analizar la correlacion llegando a la conclusion de que si p es menor a 0.05 tiene una correlacion valiosa y si es mas alto de 0.5 no tienen una correlacion muy buena   
- Se van a eliminar las siguientes columnas al no tener una buena correlacion con nuestra columna pred:    
  - patientid  
  - Insurance  
- Observacion:
  - Admission_Deposit (despues de testearlo con y sin esta columna se llego a la conclusion que no se va a eliminar ya que mejora al modelo)
 


In [21]:
#creamos nuestro df con el que vamos a trabajar la parte de ML
df_ml = df_hosp.drop(["patientid","Insurance"], axis=1)

##### normalizacion de datos


In [22]:
#aplicamos labelencoder a las columnas Age y Severity of Illness porque despues del testeo con dummi, da mejor rendimiento de esta manera
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
df_ml['Age'] = le.fit_transform(df_ml['Age'])
df_ml['Severity of Illness'] = le.fit_transform(df_ml['Severity of Illness'])

La función pd.get_dummies() de Pandas se utiliza para convertir una columna de datos categóricos en una matriz de variables dummy (también conocida como codificación one-hot)

In [23]:
df_ml = pd.get_dummies(df_ml)

In [24]:
df_ml.sample()

Available Extra Rooms in Hospital  staff_available  Age  \
244319                                  5                2    1   

        Severity of Illness  Visitors with Patient  Admission_Deposit  pred  \
244319                    2                      2        4207.851239     0   

        Department_TB & Chest disease  Department_anesthesia  \
244319                              0                      0   

        Department_gynecology  ...  gender_Other  Type of Admission_Emergency  \
244319                      1  ...             0                            1   

        Type of Admission_Trauma  Type of Admission_Urgent  \
244319                         0                         0   

        health_conditions_Asthama  health_conditions_Diabetes  \
244319                          1                           0   

        health_conditions_Heart disease  \
244319                                0   

        health_conditions_High Blood Pressure  health_conditions_None  \
244319                                      0                       0   

        health_conditions_Other  
244319                        0  

[1 rows x 39 columns]

In [25]:
#creamos nuestras variables y, x
x = df_ml.drop(columns="pred")
y = df_ml["pred"]

In [26]:
#separamos nuestros datos en entrenamiento y test
#despues de varios testeos se llego a la conclusion que en un random_state de 32 muestra su mejor performance
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.25, random_state=32)


#### PIPELINE  
Creamos nuestros pipelines compararemos la performance de los modelos y buscaremos identificar el mejor de ellos. Una vez obtenido, lo guardamos en un archivo.  
Observacion: no vamos a utilizar svm ya que su rendimiento no es optimo en nuestro computador.

In [27]:
# Importamos las librerías necesarias

from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn import tree

In [28]:
# Creamos el primer pipeline

pipe_lr = Pipeline([('scl', StandardScaler()),
			('pca', PCA(n_components=2)),
			('clf', LogisticRegression(random_state=42))])

# Creamos el segundo pipeline

pipe_dt = Pipeline([('scl', StandardScaler()),
			('pca', PCA(n_components=2)),
			('clf', tree.DecisionTreeClassifier(random_state=42))])

In [29]:
# Los guardamos en una lista

pipelines = [pipe_lr, pipe_dt]

In [30]:
# Hacemos un diccionario para fines organizativos

pipe_dict = {0: 'Regresión Logística', 1: 'Árbol de decisión'}

In [31]:
# Entrenamos

for pipe in pipelines:
	pipe.fit(x_train, y_train)

In [32]:
# Evaluamos

for idx, val in enumerate(pipelines):
	print('%s pipeline accuracy en test: %.3f' % (pipe_dict[idx], val.score(x_test, y_test)))

Regresión Logística pipeline accuracy en test: 0.622
Árbol de decisión pipeline accuracy en test: 0.629


In [33]:
# Identificamos el mejor modelo para el set de testeo

best_acc = 0.0
best_clf = 0
best_pipe = ''
for idx, val in enumerate(pipelines):
	if val.score(x_test, y_test) > best_acc:
		best_acc = val.score(x_test, y_test)
		best_pipe = val
		best_clf = idx
print('Modelo con el mejor accuracy: %s' % pipe_dict[best_clf])

Modelo con el mejor accuracy: Árbol de decisión


In [34]:
# Guardamos el pipeline en un archivo

import joblib #pickle es alternativa

joblib.dump(best_pipe, 'Mejor_pipeline.pkl', compress=1)
print('Pipeline de %s guardado a archivo' % pipe_dict[best_clf])

Pipeline de Árbol de decisión guardado a archivo


### Árbol de decisión  

Voy a utilizar el Árbol de decisión para el modelo ya que fue la recomendacion del pipeline luego de la evaluacion y es con el que mas familiarizado estoy.

In [35]:
#importamos y entrenamos el arbol de decicion
from sklearn.tree import DecisionTreeClassifier

#se usa una profundidad de 9 luego de provar varias y llegar a la conclusion que es la que mejor resultados arroja.
#se usa entropy porque lanzo un mejor recall
arbol = DecisionTreeClassifier(criterion="entropy",max_depth=16, random_state=42)
#entrenamiento
arbol.fit(x_train,y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=16, random_state=42)

In [36]:
#realizamos la prediccion
y_pred = arbol.predict(x_test)

In [37]:
#usamos confusion matrix para visualizar el desempeño de un algoritmo
#podemos observar que el modelo trabaja mucho mejor encontrando los 1 verdaderos, que los 0. 
from sklearn.metrics import confusion_matrix

cm= confusion_matrix(y_test,y_pred)
cm

array([[25295, 13250],
       [10993, 52962]], dtype=int64)

In [38]:
#verificamos las metricas de nuestro modelo
from sklearn import metrics

print("accuracy: ", metrics.accuracy_score(y_test,y_pred))
print("Recall: ", metrics.recall_score(y_test,y_pred))
print("F1score:", metrics.f1_score(y_test,y_pred))

accuracy:  0.7634829268292683
Recall:  0.8281135173168634
F1score: 0.8137546382723732


Ahora utilizamos K-fold para realizar la validación cruzada y evaluar que tan bien funciona el modelo que desarrollamos:

In [39]:
# Validación cruzada k fold
from sklearn.model_selection import KFold
kfold_validacion = KFold(10) # Acá indicamos cuantos fold queremos. En nuestro caso elegimos 10.

In [40]:
from sklearn.model_selection import cross_val_score

resultados = cross_val_score(arbol, x, y, cv = kfold_validacion)
print(resultados)
resultados.mean() # para ver el promedio de los resultados 

[0.75787805 0.75539024 0.75739024 0.75590244 0.75934146 0.76007317
 0.75863415 0.759      0.7617561  0.75421951]


0.7579585365853658

#### Conclusion:  
- El modelo arroja un accuracy de aproximadamente 70% con un recall del 88% y un F1 score del 78% encontrando en su mayoria los 1 y flaqueando en encontrar los 0.
- Actualizacion: dejando la columna "Admission_Deposit" y cambiando la columna "Age" y 'Severity of Illness' con Labelencoder el acc aumento un 6% aproximadamente, bajando un 6% el recall y aumentando el F1 score en un 3%, pero quedando mas balanceado.  
Dejando como resultado : accuracy:76%, Recall: 82.81% y F1score: 81.37%

Una vez terminado la primera etapa pasamos a la siguiente donde vamos a usar el csv de "test" para predecir y crear la culumna "pred". Antes de realizar esto tenemos que normalizar los datos borrando la mismas columnas y poniendolo en el mismo formato con el que trabajamos el arbol

In [41]:
df_test = pd.read_csv("dataset\hospitalizaciones_test.csv")

In [42]:
#dropeamos las columnas que no usaremos
df_test = df_test.drop(["patientid","Insurance"], axis=1)

In [43]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
df_test['Age'] = le.fit_transform(df_test['Age'])
df_test['Severity of Illness'] = le.fit_transform(df_test['Severity of Illness'])

In [44]:
#normalizamos los datos con dummis
df_test = pd.get_dummies(df_test)

In [45]:
#realizamos la prediccion
y_pred = arbol.predict(df_test)

In [46]:
#transformamos el array en df y le cambiamos el nombre a la columna
pred = pd.DataFrame(y_pred)
pred = pred.rename(columns={"0":"pred"})

In [49]:
#lo hacemos al df un csv y le sacamos el index
pred.to_csv("MaxiDS", index=False)